resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/undo/snapshot/IResourceSnapshot.java

```diff
@@ -0,0 +1,109 @@
+/*******************************************************************************
+ * Copyright (c) 2007, 2023 IBM Corporation and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *     Red Hat Inc - Adapted from classes in org.eclipse.ui.ide.undo and org.eclipse.ui.internal.ide.undo
+ ******************************************************************************/
+
+package org.eclipse.core.resources.undo.snapshot;
+
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+
+/**
+ * IResourceSnapshot is a lightweight snapshot that describes the common
+ * attributes of a resource to be created.
+ *
+ * @noimplement This interface is not intended to be implemented by clients.
+ * @noextend This interface is not intended to be extended by clients.
+ * @since 3.20
+ *
+ */
+public interface IResourceSnapshot {
+
+	/**
+	 * Create a resource handle that can be used to create a resource from this
+	 * resource description. This handle can be used to create the actual
+	 * resource, or to describe the creation to a resource delta factory.
+	 *
+	 * @return the resource handle that can be used to create a resource from
+	 *         this description
+	 */
+	public IResource createResourceHandle();
+
+	/**
+	 * Get the name of this resource.
+	 *
+	 * @return the name of the Resource
+	 */
+	public String getName();
+
+	/**
+	 * Create an existent resource from this resource description.
+	 *
+	 * @param monitor
+	 *            the progress monitor to use
+	 * @return a resource that has the attributes of this resource description
+	 * @throws CoreException if creation failed
+	 */
+	public IResource createResource(IProgressMonitor monitor) throws CoreException;
+
+	/**
+	 * Given a resource handle, create an actual resource with the attributes of
+	 * the receiver resource description.
+	 *
+	 * @param resource
+	 *            the resource handle
+	 * @param monitor
+	 *            the progress monitor to be used when creating the resource
+	 * @throws CoreException if creation failed
+	 */
+	public void createExistentResourceFromHandle(IResource resource,
```

Comment:
I agree it looks like this particular thing can be improved. I'm also curious about why is this method useful when we already have `createResource`. @robstryker in which case would one want to call it instead of `createResource`? What about a new `restoreResource` that would join both?